presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,271 @@
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private final Table materializedView;
+    private final Query materializedViewQuery;
+
+    private MaterializedViewInfo materializedViewInfo;
+
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query materializedViewQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            this.materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex.getMessage());
+            return node;
+        }
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        QueryBody rewrittenQueryBody = (QueryBody) process(node.getQueryBody(), context);
+        return new Query(
+                node.getWith(),
+                rewrittenQueryBody,
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        if (!node.getFrom().isPresent()) {
+            throw new IllegalStateException("Query with no From clause is not rewritable by materialized view");
+        }
+        // TODO: Handle filter containment problem https://github.com/prestodb/presto/issues/16405
+        if (materializedViewInfo.getWhereClause().isPresent() && !materializedViewInfo.getWhereClause().equals(node.getWhere())) {
+            throw new IllegalStateException("Query with no where clause is not rewritable by materialized view with where clause");
+        }
+        if (materializedViewInfo.getGroupBy().isPresent() && !node.getGroupBy().isPresent()) {
+            throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
+        }
+        // TODO: Add HAVING validation to the validator https://github.com/prestodb/presto/issues/16406
+        if (node.getHaving().isPresent()) {
+            throw new SemanticException(NOT_SUPPORTED, node, "Having clause is not supported in query optimizer");
+        }
+
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
```

Comment:
We can inline all of these.

Reply:
@jainxrohit Right now it might be fine to inline all of these, but in the future when we are dealing with table alias like this:
`SELECT b1.a, b1.b FROM base b1`
We would need to change the order of visiting querySpecification (visitFrom has priority). In that case, we would have to calculate rewrittenFrom individually instead of following the order.